### PR TITLE
Validate schemas after compilation, closes #2085

### LINKED ISSUE #2085: Misleading error when Preloading associations defined with aliases
### Environment

* Elixir version: `Elixir 1.4.4`
* Database and version: `PostgreSQL 9.6.3`
* Ecto version: `phoenix_ecto 3.2.3`
* Database adapter and version: `postgrex 0.13.2`
* Operating system:  `OS X 10.12.4`

### Error

When preloading a struct through an association that is not defined with it's full module path _and_ not aliased in the file calling `preload`, you get a `module does not exist` error.

The first part of the error was fairly misleading and it took quite a bit of digging to solve, once I'd found the following gist explaining a solution: https://gist.github.com/diogovk/295e07380af4dc36b520

I'm fairly new to Elixir, so I'm not sure if this is something that is even possible to detect and produce a more meaningful error. Alternatively it may be helpful to document this behaviour in the docs section on associations.

Code: (slightly cut down example of the problem, I'm not sure if it will reproduce the error)
```elixir
# something/field.ex
defmodule Project.Something.Field do
  use Ecto.Schema
  import Ecto.Changeset
  alias Project.Something.Field
  alias Project.Something.Form
  
  schema "fields" do
    belongs_to :form, Form
  end
end

# something/form.ex
defmodule Project.Something.Form do
  use Ecto.Schema
  import Ecto.Changeset
  alias Project.Something.Form
  alias Project.Something.Field

  schema "forms" do
    has_many :fields, Field
  end
end

# something/load.ex
alias Project.Something.Form
def get_form!(id) do
  Repo.get!(Form, id)
  |> Repo.preload([:fields])
end
```

Stack trace when calling `get_form`
```elixir
** (exit) an exception was raised:
    ** (Protocol.UndefinedError) protocol Ecto.Queryable not implemented for Field, the given module does not exist
        (ecto) lib/ecto/queryable.ex:37: Ecto.Queryable.Atom.to_query/1
        (ecto) lib/ecto/query/builder/from.ex:97: Ecto.Query.Builder.From.apply/2
        (ecto) lib/ecto/association.ex:495: Ecto.Association.Has.assoc_query/3
        (ecto) lib/ecto/repo/preloader.ex:150: Ecto.Repo.Preloader.fetch_query/8
        (ecto) lib/ecto/repo/preloader.ex:111: Ecto.Repo.Preloader.preload_assoc/10
        (elixir) lib/enum.ex:1229: Enum."-map/2-lists^map/1-0-"/2
        (ecto) lib/ecto/repo/preloader.ex:63: Ecto.Repo.Preloader.preload_each/4
        (ecto) lib/ecto/repo/preloader.ex:113: Ecto.Repo.Preloader.preload_assoc/10
        (elixir) lib/task/supervised.ex:85: Task.Supervised.do_apply/2
        (elixir) lib/task/supervised.ex:36: Task.Supervised.reply/5
        (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
```

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

File: lib/ecto/association.ex [hunk 1]
```diff
@@ -417,6 +425,20 @@ defmodule Ecto.Association.Has do
...
 
   @doc false
+  def after_compile_validation(%{queryable: queryable}) do
+    cond do
+      not is_atom(queryable) ->
+        :ok
+      not Code.ensure_compiled?(queryable) ->
+        {:error, "associated schema #{inspect queryable} does not exist"}
+      not function_exported?(queryable, :__schema__, 2) ->
+        {:error, "associated module #{inspect queryable} is not an Ecto schema"}
```
Comment (ericmj) [inline lib/ecto/association.ex]: Should we check other things, like matching foreign key fields?

File: lib/ecto/association.ex [hunk 2]
```diff
@@ -807,6 +848,26 @@ defmodule Ecto.Association.ManyToMany do
...
+      not is_atom(join_through) ->
+        :ok
+      not Code.ensure_compiled?(queryable) ->
+        {:error, ":join_through schema #{inspect join_through} does not exist"}
+      not function_exported?(queryable, :__schema__, 2) ->
```
Comment (ericmj) [inline lib/ecto/association.ex]: join_through instead of queryable?

File: lib/ecto/association.ex (hunk 1 above)
Comment (josevalim) [inline lib/ecto/association.ex]: @ericmj stop giving me work to do :P

File: lib/ecto/association.ex (hunk 2 above)
Comment (josevalim) [inline lib/ecto/association.ex]: Great catch!

File: lib/ecto/association.ex (hunk 1 above)
Comment (josevalim) [inline lib/ecto/association.ex]: Done in a later commit.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by xLegoz** [comment]
Thanks for tackling this on quickly! It's awesome how simple and clear the validation is :). You all do a really awesome job with Elixir et. all!

**Pull request merged**